splitgraph/splitgraph.com
# Merge long-running branch `canary` into default branch `master` with some known regressions

This PR merges all the changes from the long-running pull request `canary`
into the `master` default branch of this repository. The goal is to reduce
cognitive overhead, especially when working with this repository as a submodule.

The current `master` branch will be pushed to a protected branch `stable`
(which isn't technically necessary since there is a tag for every master branch).

The CI step that cuts a release on merge to `master` should be temporarily disabled
prior to merge. See below for details.

## Breaking Workflow Changes

Minimal workflow changes are required – just use `master` where
you were previously using `canary`.

- Open future pull requests into `master`, not `canary`
- If you are currently working on a branch with an ancestor
on `canary` (i.e., you created the branch from `canary`),
then the next time you `rebase` or `merge`, you should do
`git rebase master` or `git merge master`, the same as you would
for `canary`. 

## Background - Why the split?

This repository contains a few partially distinct packages/resources:

- `content`: Raw `.mdx` content that is both manually and automatically generated
  - The auto-generated content comes from running the `updateDocs` script
    in `content-scripts`, which pulls documentation from the `splitgraph/splitgraph`
    repository and converts it to `.mdx` format
- `docs`: The Next.js application that creates a static bundle from the `content` directory
- `(t)design` and others: UI components / themes that the public `docs` website shares with
the internal `catalog` website (i.e., both import from `tdesign`)

Originally, the main goal of the `canary` branch was to drop `theme-ui`
in favor of `material-ui`, which meant a lot of changes touching a lot
of files. We wanted to keep the `master` branch stable for the deployment
process. This `canary` branch is not yet as stable as `master`, but even
`master` will fail builds now due to third-party Vercel deprecating the
CLI used in the build process. At this point it will be easier to bring
`canary` up-to-date.

## What's in this PR

There is quite a bit in this PR that's probably not even worth listing
exhaustively at this point, as it's effectively become the default branch
anyway. At a high level:

- Bring major dependencies up to date 
- Migrate from `theme-ui` and `rebass` to `material-ui` (using v5 alpha,
  still not latest though see #120)
- Migrate to TypeScript project references
- Setup Yarn to work within submodule with a parent `yarn.lock`, or in isolation here,
  with `yarn-public-workspace.lock`
- Add a bunch of new components to `@splitgraph/design`, mostly used in auth
  and settings pages
- Add tooling: `pre-commit`, `madge` for static analysis (check circular deps),
  some hacky linting (just using bash, no real linter atm)
- Add Storybook (currently temporarily broken due to
  WIP theming changes, see #124)
- Resolve *most* of the regressions in `docs` after MUI upgrade (it builds), but
  deprioritize many of them since we were focused on shipping `catalog` (the app)
- Mostly migrate JS files from `design` -> `tdesign` (as mentioned above, still some left)

## Still Broken

This branch still has some regressions from the current `master`.

### Preview deployments on every commit

Previously, every commit to any branch would build the static bundle, and upload
it to a preview URL hosted by Vercel. This is currently broken for two distinct reasons:

1. Vercel has deprecated the `now` CLI that we used to deploy the bundle
1. Vercel is designed to run the build scripts on their machines, but we
were running them in GitHub actions, then uploading the entire bundle
to Vercel. Due to the size of the bundle from the autogenerated content,
this caused rate limiting errors that caused builds to fail. The original
reasoning for this setup was to workaround a lack of Yarn 2 compatibility.

### Release bundles on every commit to `master`

In theory, deployment of the `docs` repository happens only from the `master` branch.
Every commit to `master` triggers a GitHub action to build the `docs` website
into a static bundle and upload it as a release to GitHub tagged with the current
date. Then, our production OpenResty server can download the latest bundle and
extract it into its static files directory to be served as static content.

This build/release process is currently blocked for one primary reason:

1. The `docs` website itself (landing page, documentation, and some inner pages)
has multiple styling regressions after the upgrade to MUI. This is due to
focusing on `catalog` compatibility with MUI to ship that `canary` a few weeks ago.

Since we are working on the new website in a new repository, it doesn't make sense
to prioritize fixing the landing page that's about to be deprecated. We will be
addressing the documentation soon. We also need to switch to something like
`next-mdx-remote` and SSG instead of using Babel macros (which were the best option
when we made them but are now the worst option).

## Heads Up

A few things to watch out for / keep in mind...

- This branch directly changed the imports in many of the `content` files, 
including those that are autogenerated (e.g. deleting
`import "@splitgraph/design/css/sphinxtheme.css";`). The `updateDocs`
script will need to be updated accordingly before running it again.

- The production website at `splitgraph.com` hotlinks images and resources
using the `raw.githubusercontent.com` URL with the `master` branch hardcoded.
Therefore it would be possible for this PR to break production without actually
deploying anywhere, if e.g. it deleted or moved any of those files. However,
AFAICT it does not do this, so it's safe to at least merge this to `master`.

- There are still a few WIP tasks in the migration, e.g. we mostly moved
all of the JS files from `design` into `tdesign`, but we haven't renamed
`tdesign` to `design` yet (there are still a few stragglers).
